stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,241 @@
+                "{vendor_id}:{device_id}. Does this correspond to a network "
+                "card?".format(vendor_id=vendor_id, device_id=device_id))
+        interfaces.append(interface_name)
+    return dict(map(lambda x: (
+        x, _get_ethtool_field(_get_ethtool_output(x), "firmware-version")),
+        interfaces))
+
+
+class SystemNICHardwareManager(hardware.HardwareManager):
+    """Checks firmware version for a given network card"""
+
+    HARDWARE_MANAGER_VERSION = '1.0'
+
+    def evaluate_hardware_support(self):
+        """Declare whether the system is supported by this manager.
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool, lspci
+        return hardware.HardwareSupport.SERVICE_PROVIDER
+
+    def get_clean_steps(self, node, ports):
+        """Get a list of clean steps with priority.
+
+        :param node: The node object as provided by Ironic.
+        :param ports: Port objects as provided by Ironic.
+        :returns: A list of cleaning steps, as a list of dicts.
+        """
+        return [{'step': 'verify_nic_firmware',
+                 'priority': 90,
+                 'interface': 'deploy',
+                 'reboot_requested': False,
+                 'abortable': True}]
+
+    def _get_lspci_output(*args, **kwargs):
+        return _get_lspci_output(*args, **kwargs)
+
+    def _pci_addr_to_net_interface(*args, **kwargs):
+        return _pci_addr_to_net_interface(*args, **kwargs)
+
+    def _get_ethtool_output(*args, **kwargs):
+        return _get_ethtool_output(*args, **kwargs)
+
+    def _get_nic_firmware_versions(self, vendor_id, device_id):
+        # there might be multiple identical cards, we must check them all
+        devices = _parse_lspci_output(
+            self._get_lspci_output(vendor_id, device_id))
+        interfaces = []
+        for device in devices:
+            interface_name = self._pci_addr_to_net_interface(device['Slot'])
+            if interface_name is None:
+                raise errors.CleaningError(
+                    "Could not network determine interface name. The pci_id "
+                    "was: {vendor_id}:{device_id}. Does this correspond to a "
+                    "network card?"
+                    .format(vendor_id=vendor_id, device_id=device_id))
+            interfaces.append(interface_name)
+        return dict(map(lambda x: (x, _get_ethtool_field(
+            self._get_ethtool_output(x), "firmware-version")), interfaces))
+
+    def verify_nic_firmware(self, node, ports):
+        """
+        Verify all firmware versions specified by the nic_firmware property.
+        """
+
+        if "extra" not in node or "nic_firmware" not in node["extra"]:
+            LOG.warning(
+                'NIC firmware property has not been set. No firmware will be '
+                'verified')
+            return True
```

Comment:
Seems sensible. I guess the alternative could be to fail cleaning if no firmware is set, unless a flag ```skip_nic_firmware_check``` is set. For environments where you want to check that all firmware is as expected that might help to enforce good practice. Anyone else have an opinion?

Reply:
I thought the absence of that data structure would indicate that you didn't want to check the firmware, but we could add a flag if you think that would be better.

Reply:
Yeah - I guess it's a question of do we want to force people to opt out by explicitly having to set a flag, or let them opt in.

Reply:
what did we do with system_bios?

Reply:
We have a separate flag for disabling the check - I'll change so this is consistent.